src/random.h

```diff
@@ -25,6 +26,29 @@ int GetRandInt(int nMax);
 uint256 GetRandHash();
 
 /**
+ * Rearranges the elements in the range [first,first+len) randomly, assuming
+ * that gen is a uniform random number generator. Follows the same algorithm as
+ * std::shuffle in C++11 (a Durstenfeld shuffle).
+ *
+ * The elements in the range [mapFirst,mapFirst+len) are rearranged according to
+ * the same permutation, enabling the permutation to be tracked by the caller.
+ *
+ * gen takes an integer n and produces a uniform random output in [0,n).
+ */
+template <typename RandomAccessIterator, typename MapRandomAccessIterator>
+void MappedShuffle(RandomAccessIterator first,
+                   MapRandomAccessIterator mapFirst,
+                   size_t len,
```

Comment:
Or just `last-first`, which is what the equivalence code in `std::shuffle` uses. But that means we have to pass in both `begin()` and `end()` for both collections (which is why @daira and I opted at the time for `len`). If you prefer that style, I can change it.